File: algo/739.py
from typing import List

class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        stack: List[int] = []
        answer = [0] * len(temperatures)

        for i, temperature in enumerate(temperatures):
            while stack and temperature > temperatures[stack[-1]]:
                last = stack.pop()
                answer[last] = i - last
            stack.append(i)

        return answer


d = Solution()
temperatures = [73,74,75,71,69,72,76,73]
print(d.dailyTemperatures(temperatures))
# Input: temperatures = [73,74,75,71,69,72,76,73]
# Output: [1,1,4,2,1,1,0,0]

temperatures = [30,40,50,60]
print(d.dailyTemperatures(temperatures))
# Input: temperatures = [30,40,50,60]
# Output: [1,1,1,0]

temperatures = [30,60,90]
print(d.dailyTemperatures(temperatures))
# Input: temperatures = [30,60,90]
# Output: [1,1,0]
 

File: algo/739.kt
class Solution {
    fun dailyTemperatures(temperatures: IntArray): IntArray {
        val stack = mutableListOf(0)
        val result = IntArray(temperatures.size)
        for (i in 1..temperatures.size - 1) {
            while (stack.isNotEmpty() && temperatures[stack[stack.lastIndex]] < temperatures[i]) {
                val top = stack[stack.lastIndex]
                result[top] = i - top
                stack.removeAt(stack.lastIndex)
            }
            stack.add(i)
        }
        return result
    }
}

//val temperatures = intArrayOf(73, 74, 75, 71, 69, 72, 76, 73)
//val temperatures = intArrayOf(30, 40, 50, 60)
val temperatures = intArrayOf(30, 60, 90)
println(Solution().dailyTemperatures(temperatures).contentToString())
